service/src/ConstConfigIOGroup.cpp

```diff
@@ -0,0 +1,451 @@
+    {
+        return M_PLUGIN_NAME;
+    }
+
+    std::unique_ptr<IOGroup> ConstConfigIOGroup::make_plugin(void)
+    {
+        return geopm::make_unique<ConstConfigIOGroup>();
+    }
+
+    void ConstConfigIOGroup::save_control(const std::string &save_path)
+    {
+    }
+
+    void ConstConfigIOGroup::restore_control(const std::string &save_path)
+    {
+    }
+
+    void ConstConfigIOGroup::parse_config_json(const std::string &config)
+    {
+        Json root = construct_config_json_obj(config);
+        check_json_root(root);
+
+        auto signals = root.object_items();
+        for (const auto &signal : signals) {
+            check_json_signal(signal);
+            std::string name = M_SIGNAL_PREFIX + signal.first;
+
+            auto properties = signal.second.object_items();
+            check_json_signal_properties(signal.first, properties);
+
+            int units = IOGroup::string_to_units(
+                    properties["units"].string_value());
+            int domain_type = PlatformTopo::domain_name_to_type(
+                    properties["domain"].string_value());
+            auto agg_func = Agg::name_to_function(
+                    properties["aggregation"].string_value());
+
+            auto json_values = properties["values"].array_items();
+            std::vector<double> values;
+            for (const auto &val : json_values) {
+                check_json_array_value(val);
+                values.push_back(val.number_value());
+            }
+
+            std::string description = properties["description"].string_value();
+            // TODO: check for duplicate signals. At the moment, we're using
+            // json11 to parse JSON strings, which handles duplicate entries
+            // by taking the latest entry encountered.
+            m_signal_available[name] =
+                std::make_shared<m_signal_info_s>(
+                    m_signal_info_s {
+                        .units = units,
+                        .domain = domain_type,
+                        .agg_function = agg_func,
+                        .description = description,
+                        .values = values});
+        }
+    }
+
+    Json ConstConfigIOGroup::construct_config_json_obj(
+        const std::string &config)
+    {
+        std::string err;
+        Json root = Json::parse(config, err);
+        if (!err.empty()) {
+            throw Exception("ConstConfigIOGroup::parse_config_json(): "
+                            "detected a malformed JSON string",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        return root;
+    }
+
+    void ConstConfigIOGroup::check_json_root(const Json &root)
+    {
+        if (!root.is_object()) {
+            throw Exception("ConstConfigIOGroup::parse_config_json(): "
+                            "the root must be an object",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+    }
```

Comment:
I don't see the benefit of having these methods separated out from ```parse_config_json()```.  Take a look at some of the other places in the code base we parse json:
    ```FilePolicy::parse_json()```
    ```SaveControlImp::settings()```

Those both have code of the form:
```if (!err.empty() || !root.is_object()) {```

If you agree, please combine all of these error checks into a more condensed form, and put the code in ```parse_config_json()```.

Reply:
I wanted to separate the error checking from the actual parsing to make the parsing code easier to read (by removing the clutter of all the error checking, and keeping the function short and focused on the parsing).

I originally looked at MSRIOGroup's parse method and had a bit of a hard time following it because it's somewhat lengthy, and has some clutter as a result of the checks. A bit more distracting is that it has some of the checks inline while it farms out some checks to other functions.

I looked at FilePolicy's parse method and I do see it has its error checking inline. However, its parsing logic is simple enough that the method is pretty short and easy to read, in which case the error checking doesn't cause a distraction.

I generally try to keep functions reasonably short and focused on one purpose to keep them easy to read, but I know some of this is subjective. In this case I would prefer to keep the parse method as-is, but if you feel strongly that it's a deviation on an important pattern being followed in the codebase and feel it makes the function hard to read, then I'm happy to make the suggested change.

Reply:
Update: Daniel has suggested looking into using json11's ``has_shape()`` method to simplify the checking. Looking at it, I see I can leverage it to at least check required fields and their types. I'll do a pass at improving the checking with that.

Reply:
Update: simplified the parsing method to an extent by reducing the number of "check" functions by leveraging the ``has_shape()`` method.